packages/user/DiffAdjust/service/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+            self.check_sender_has_nft();
+            self.floor_difficulty = floor_difficulty;
+            self.active_difficulty = self.active_difficulty.max(floor_difficulty);
+            self.save();
+        }
+
+        pub fn delete(&self) {
+            self.check_sender_has_nft();
+            RateLimitTable::read_write().remove(&self);
+        }
+
+        fn save(&self) {
+            RateLimitTable::read_write()
+                .put(&self)
+                .expect("failed to save");
+        }
+
+        pub fn get(nft_id: u32) -> Option<Self> {
+            RateLimitTable::read().get_index_pk().get(&nft_id)
+        }
+
+        pub fn get_assert(nft_id: u32) -> Self {
+            check_some(Self::get(nft_id), "rate limit of NFT ID does not exist")
+        }
+    }
+
+    #[ComplexObject]
+    impl RateLimit {
+        pub async fn admin(&self) -> AccountNumber {
+            Nft::call().getNft(self.nft_id).owner
+        }
+    }
+}
+
+#[psibase::service(name = "diff-adjust", tables = "tables")]
+pub mod service {
+    use crate::tables::RateLimit;
+
+    /// Creates a new Rate limit
+    ///
+    /// # Arguments
+    /// * `initial_difficulty` - Sets initial difficulty
+    /// * `window_seconds` - Seconds duration before decay occurs
+    /// * `target_min` - Minimum rate limit target
+    /// * `target_max` - Maximum rate limit target
+    /// * `floor_difficulty` - Minimum price
+    /// * `percent_change` - Percent to increment / decrement, 50000 = 5%
+    #[action]
+    fn create(
+        initial_difficulty: u64,
+        window_seconds: u32,
+        target_min: u32,
+        target_max: u32,
+        floor_difficulty: u64,
+        percent_change: u32,
+    ) -> u32 {
+        RateLimit::add(
+            initial_difficulty,
+            window_seconds,
+            target_min,
+            target_max,
+            floor_difficulty,
+            percent_change,
+        )
+        .nft_id
+    }
+
+    /// Get RateLimit price
+    ///
+    /// # Arguments
+    /// * `nft_id` - RateLimit / NFT ID
+    ///
+    /// # Returns
+    /// Price of RateLimit
+    #[action]
+    fn get_price(nft_id: u32) -> u64 {
+        RateLimit::get_assert(nft_id).check_difficulty_decrease()
+    }
+
+    /// Increment RateLimit instance, potentially increasing RateLimit
+    ///
+    /// * Requires holding administration NFT.
+    ///
+    /// # Arguments
+    /// * `nft_id` - RateLimit / NFT ID
+    /// * `amount` - Amount to increment the counter by
+    #[action]
+    fn increment(nft_id: u32, amount: u32) -> u64 {
+        RateLimit::get_assert(nft_id).increment(amount)
+    }
+
+    /// Update targets
+    ///
+    /// * Requires holding administration NFT.
+    ///
+    /// # Arguments
+    /// * `nft_id` - RateLimit / NFT ID
+    /// * `target_min` - Minimum target difficulty
+    /// * `target_max` - Maximum target difficulty
+    #[action]
+    fn up_targets(nft_id: u32, target_min: u32, target_max: u32) {
```

Comment:
`up` might not be the best prefix. My first thought when I saw it was that it adds to the current values rather than setting them.